### PR TITLE
Fix behaviour of inputs editing in block settings

### LINKED ISSUE #1090: Feature Request (Tool-Setting): add option to not delete selected block on backspace or enter keydown
## Scenario

A block's tool setting opens a modal (or any other new view outside of editorjs). Whenever the user hits backspace _or_ enter in the new view though, the block - which is still selected (`.ce-block--selected`) - gets deleted.

There is currently _no way_ to remove the selection from the block, when a tune was toggled/clicked.

## Example

Screen Capture on 2020-04-01 at 16-40-17

## Example Codepen

https://codepen.io/natterstefan/pen/PoqLVJr?editors=0110

## Research

There is a `enableLineBreaks` setting, managing how blocks behave when `enter` is pressed. In [the example](https://codepen.io/natterstefan/pen/PoqLVJr?editors=0110) it still looks like this is not working properly.

https://github.com/codex-team/editor.js/blob/be6f9b78f265c3d553c799c99e59c9916bf29e17/src/components/modules/blockEvents.ts#L257-L263

Same goes for `backspace`, but here we never reach this part of the code:

https://github.com/codex-team/editor.js/blob/be6f9b78f265c3d553c799c99e59c9916bf29e17/src/components/modules/blockEvents.ts#L353-L361

Because the block is still selected and [results to true in this if](https://github.com/codex-team/editor.js/blob/be6f9b78f265c3d553c799c99e59c9916bf29e17/src/components/modules/blockEvents.ts#L327-L338
).

## Feature request

Enable developers to unselect a block when a tune was toggled. Means, a new method is added to the blocks API (https://editorjs.io/api#blocks).

## Current Workaround (not 100% tested tough)

When the toggle is clicked, we remove the [selected class](https://github.com/codex-team/editor.js/blob/be6f9b78f265c3d553c799c99e59c9916bf29e17/src/components/block.ts#L67) from the block. This should help us to pass the if`s in the keydown handlers.

But this is kinda hacky IMHO.

## Additional questions

- is there really a global active `keydown` listener on the document ([here](https://github.com/codex-team/editor.js/blob/be6f9b78f265c3d553c799c99e59c9916bf29e17/src/components/modules/ui.ts#L310))? Because there is another [`backspacePressed` handler](https://github.com/codex-team/editor.js/blob/be6f9b78f265c3d553c799c99e59c9916bf29e17/src/components/modules/ui.ts#L390).

## (Possible) related Issues

- https://github.com/codex-team/editor.js/issues/810

### PR DESCRIPTION
**Pull request by gohabereg**
Resolves #1090

### CODE REVIEW

**Comment by gohabereg** [comment]
Hi @natterstefan @m-alzam,

This PR should resolve problems with `enter` and `backspace` keys. Could you please check it in your use-cases?

File: src/components/domIterator.ts [hunk 1]
```diff
@@ -156,7 +156,8 @@ export default class DomIterator {
       focusedButtonIndex = (this.items.length + focusedButtonIndex - 1) % this.items.length;
     }
 
-    if (Dom.isNativeInput(this.items[focusedButtonIndex])) {
+    if (Dom.canSetCaret(this.items[focusedButtonIndex])) {
+
```
Comment (neSpecc) [inline src/components/domIterator.ts:160]: there can be contenteditable element, so you need to use custom `focus()` method

File: src/components/flipper.ts [hunk 2]
```diff
@@ -190,7 +213,7 @@ export default class Flipper {
       _.keyCodes.ENTER,
     ];
 
-    if (this.allowArrows) {
+    if (this.allowArrows && this.iterator.currentItem !== document.activeElement) {
```
Comment (neSpecc) [inline src/components/flipper.ts:216]: need to describe this case or move the second condition to the variable

File: src/components/flipper.ts [hunk 3]
```diff
@@ -199,11 +222,7 @@ export default class Flipper {
       );
     }
 
-    if (!this.activated || handlingKeyCodeList.indexOf(event.keyCode) === -1) {
-      return false;
-    }
-
-    return true;
+    return (this.activated && handlingKeyCodeList.indexOf(event.keyCode) !== -1);
```
Comment (neSpecc) [inline src/components/flipper.ts:225]: extra parentheses

File: src/components/modules/blockSelection.ts [hunk 4]
```diff
@@ -188,17 +188,20 @@ export default class BlockSelection extends Module {
...
 
+    const isKeyboard = reason && (reason instanceof KeyboardEvent);
+    const isPrintableKey = isKeyboard && _.isPrintableKey((reason as KeyboardEvent).keyCode);
+
     /**
      * If reason caused clear of the selection was printable key and any block is selected,
      * remove selected blocks and insert pressed key
      */
-    if (this.anyBlockSelected && reason && reason instanceof KeyboardEvent && _.isPrintableKey(reason.keyCode)) {
+    if (this.anyBlockSelected && isKeyboard && isPrintableKey && !SelectionUtils.exists) {
```
Comment (neSpecc) [inline src/components/modules/blockSelection.ts:198]: `SelectionUtils.exists` - bad naming. Exists what?

File: src/components/modules/toolbar/blockSettings.ts [hunk 5]
```diff
@@ -158,7 +176,10 @@ export default class BlockSettings extends Module {
       return this.buttons;
     }
 
-    const toolSettings = this.nodes.toolSettings.querySelectorAll(`.${this.Editor.StylesAPI.classes.settingsButton}`);
+    const toolSettings = this.nodes.toolSettings.querySelectorAll(
+      // Select buttons and inputs
+      `.${StylesAPI.classes.settingsButton}, .${StylesAPI.classes.input}`
```
Comment (neSpecc) [inline src/components/modules/toolbar/blockSettings.ts:181]: why not use `$.findAllInputs()` here?

File: src/components/modules/toolbar/blockSettings.ts [hunk 6]
```diff
@@ -195,7 +216,18 @@ export default class BlockSettings extends Module {
   private enableFlipper(): void {
     this.flipper = new Flipper({
       focusedItemClass: this.CSS.focusedButton,
-      activateCallback: () => {
+      activateCallback: (activatedItem) => {
```
Comment (neSpecc) [inline src/components/modules/toolbar/blockSettings.ts:219]: the parameter is not described. maybe name it `focusedItem`?

File: src/components/modules/ui.ts [hunk 7]
```diff
@@ -410,7 +416,7 @@ export default class UI extends Module {
   private backspacePressed(event: KeyboardEvent): void {
     const { BlockManager, BlockSelection, Caret } = this.Editor;
 
-    if (BlockSelection.anyBlockSelected) {
+    if (BlockSelection.anyBlockSelected && !Selection.exists) {
```
Comment (neSpecc) [inline src/components/modules/ui.ts:419]: need to describe why the second part was added

File: src/components/modules/ui.ts [hunk 8]
```diff
@@ -438,7 +470,7 @@ export default class UI extends Module {
     const { BlockManager, BlockSelection, Caret } = this.Editor;
     const hasPointerToBlock = BlockManager.currentBlockIndex >= 0;
 
-    if (BlockSelection.anyBlockSelected) {
+    if (BlockSelection.anyBlockSelected && !Selection.exists) {
```
Comment (neSpecc) [inline src/components/modules/ui.ts:473]: need to describe why the second part was added

File: src/components/modules/toolbar/blockSettings.ts [hunk 9]
```diff
@@ -195,7 +216,18 @@ export default class BlockSettings extends Module {
...
+
+        /**
+         * If activated item is editable element, restore selection to block and flip to the next item
```
Comment (neSpecc) [inline src/components/modules/toolbar/blockSettings.ts:222]: can't understand this moment. Why do we need to restore selection to block, if, for example, an input at block settings is focused

Comment (gohabereg) [reply src/components/modules/toolbar/blockSettings.ts:222]: I thought about this scenario:
1. Open Block settings via tab
2. Type something to an editable element there
3. Press enter

What should happened after that?
For sure, Tune should do something with the typed text, but what about settings block. Should it be closed? At the moment caret is returned to the Block and next Tune is selected

File: src/components/dom.ts [hunk 10]
```diff
@@ -573,4 +599,31 @@ export default class Dom {
   public static isAnchor(element: Element): element is HTMLAnchorElement {
     return element.tagName.toLowerCase() === 'a';
   }
+
+  /**
+   * Set focus to contenteditable or native input element
+   *
+   * @param element - element where to set focus
+   */
+  public static focus(element: HTMLElement): void {
```
Comment (neSpecc) [inline src/components/dom.ts:608]: can we use this method in module/caret@set ?

Comment (neSpecc) [reply src/components/dom.ts:608]: and maybe it's better to place it to the selectionUtis?

**Comment by m-alzam** [comment]
Hi @gohabereg, I'll try test your patch tomorrow with my plugin. Thanks for it. Should I abandon my PR?

**Comment by m-alzam** [comment]
Hi @gohabereg: the patch works fine for me. Thanks!

File: src/components/modules/toolbar/blockSettings.ts [hunk 11]
```diff
@@ -130,6 +141,15 @@ export default class BlockSettings extends Module {
   public close(): void {
     this.nodes.wrapper.classList.remove(this.CSS.wrapperOpened);
 
+    /**
+     * Restore selection to Block
```
Comment (neSpecc) [inline src/components/modules/toolbar/blockSettings.ts:145]: need to describe current condition: why we need to check if the selection is at editor

File: src/components/modules/toolbar/blockSettings.ts (hunk 5 above)
Comment (neSpecc) [reply src/components/modules/toolbar/blockSettings.ts:185]: can we write a method for accepting `cdx-buttons` and every-inputs with saving their order at dom?

File: src/components/selection.ts [hunk 12]
```diff
@@ -266,6 +275,37 @@ export default class SelectionUtils {
     return window.getSelection();
   }
 
+  /**
+   * Set focus to contenteditable or native input element
+   *
+   * @param element - element where to set focus
+   * @param offset - offset of cursor
+   */
```
Comment (neSpecc) [inline src/components/selection.ts:283]: describe what it returns

**Pull request merged**